lib/unifex/code_generator/base_types/list.ex

```diff
@@ -27,66 +59,99 @@ defmodule Unifex.CodeGenerator.BaseTypes.List do
-      ]
-    end
-
     @impl BaseType
     def generate_arg_parse(arg, var_name, ctx) do
       elem_name = :"#{var_name}[i]"
       len_var_name = "#{var_name}_length"
+      native_type = BaseType.generate_native_type(ctx.subtype, ctx.generator)
+      %{subtype: subtype, postproc_fun: postproc_fun, generator: generator} = ctx
 
       ~g"""
       ({
       int get_list_length_result = enif_get_list_length(env, #{arg}, &#{len_var_name});
       if(get_list_length_result){
-        #{var_name} = enif_alloc(
-          sizeof(#{BaseType.generate_native_type(ctx.subtype, ctx.generator)}) * #{len_var_name});
+        #{var_name} = enif_alloc(sizeof(#{native_type}) * #{len_var_name});
 
         for(unsigned int i = 0; i < #{len_var_name}; i++) {
-          #{BaseType.generate_initialization(ctx.subtype, elem_name, ctx.generator)}
+          #{BaseType.generate_initialization(subtype, elem_name, generator)}
         }
 
         ERL_NIF_TERM list = #{arg};
         for(unsigned int i = 0; i < #{len_var_name}; i++) {
           ERL_NIF_TERM elem;
           enif_get_list_cell(env, list, &elem, &list);
-          #{
-        BaseType.generate_arg_parse(
-          ctx.subtype,
-          elem_name,
-          ~g<elem>,
-          ctx.postproc_fun,
-          ctx.generator
-        )
-      }
+          #{BaseType.generate_arg_parse(subtype, elem_name, ~g<elem>, postproc_fun, generator)}
         }
       }
       get_list_length_result;
       })
       """
     end
+  end
+
+  defmodule CNode do
+    @moduledoc false
+    use Unifex.CodeGenerator.BaseType
+    alias Unifex.CodeGenerator.BaseType
 
     @impl BaseType
-    def generate_initialization(name, _ctx) do
-      ~g<#{name} = NULL;>
+    def generate_arg_serialize(name, ctx) do
+      ~g"""
+      ({
+        ei_x_encode_list_header(out_buff, #{name}_length);
+        for(unsigned int i = 0; i < #{name}_length; i++) {
+          #{BaseType.generate_arg_serialize(ctx.subtype, :"#{name}[i]", ctx.generator)}
+        }
+        ei_x_encode_empty_list(out_buff);
+      });
+      """
     end
 
     @impl BaseType
-    def generate_destruction(name, ctx) do
+    def generate_arg_parse(arg, var_name, ctx) do
+      elem_name = :"#{var_name}[i]"
+      len_var_name = "#{var_name}_length"
+      native_type = BaseType.generate_native_type(ctx.subtype, ctx.generator)
+      %{subtype: subtype, postproc_fun: postproc_fun, generator: generator} = ctx
+
       ~g"""
-      if(#{name} != NULL) {
-        for(unsigned int i = 0; i < #{name}_length; i++) {
-          #{BaseType.generate_destruction(ctx.subtype, :"#{name}[i]", ctx.generator)}
+      ({
+        int res = 1;
+        int type;
+        int size;
+
+        ei_get_type(#{arg}->buff, #{arg}->index, &type, &size);
+        #{len_var_name} = (unsigned int) size;
+
+        UnifexCNodeInBuff *unifex_buff = (UnifexCNodeInBuff *)malloc(sizeof(UnifexCNodeInBuff));
+        unifex_buff->index = (int *)malloc(sizeof(int));
```

Comment:
Both of these variables can be allocated on the stack. That way you wouldn't need to free unifex_buff and unifex_buff->index (the latter seems missing).

Reply:
I need to pass it  to `BaseType.generate_arg_parse()` which require pointer and this was the reason. I mean `BaseType.generate_arg_parse()` generates code with `->` not with `.`.

Reply:
You can always obtain a pointer with `&variable`, it will be valid till the function returns